Scripts/Playthrough.cs

```diff
@@ -374,6 +341,208 @@ private bool IsConnected()
+            {
+                Logger.Log($"Received `stop-typing` event: {JsonConvert.SerializeObject(message)}");
+                OnStopTyping?.Invoke(message);
+            });
+
+            _room.OnMessage<Events.ProblemEvent>("problem", (message) =>
+            {
+                Logger.LogWarning($"Received `problem` event: {JsonConvert.SerializeObject(message)}");
+            });
+
+            _room.OnMessage<string>("pong", (message) =>
+            {
+                _pingCount = 0;
+                OnPingSuccess?.Invoke();
+                SetConnectionState(ConnectionState.Connected);
+            });
+
+            _room.OnLeave += (code) =>
+            {
+                Logger.Log($"Connection closed. Attempting to reconnect to Playthrough.");
+                MainThreadConsumer.Instance?.StartCoroutine(TryToReconnect());
+            };
+        }
+
+        private void OnReady()
+        {
+            SetConnectionState(ConnectionState.Connected);
+
+            _onReadyCallback?.Invoke();
+
+            _calledByDisconnect = false;
+            _reconnectionTryCount = 0;
+
+            Task.Run(FirePing);
+        }
+
+        private IEnumerator TryToReconnect()
+        {
+            Logger.Log("Trying to reconnect.");
+            if (_calledByDisconnect)
+            {
+                SetConnectionState(ConnectionState.Disconnected);
+                yield break;
+            }
+
+            // cannot allow multiple reconnection coroutines to run concurrently
+            if (IsReconnecting())
+            {
+                yield break;
+            }
+
+            SetConnectionState(ConnectionState.Reconnecting);
+
+            MainThreadConsumer.Instance.StartCoroutine(Reconnect());
+        }
+
+        private bool IsReconnecting()
+        {
+            return _connectionState == ConnectionState.Reconnecting;
+        }
+
+        private IEnumerator Reconnect()
+        {
+            ColyseusRoom<object> room = null;
+
+            while (room == null)
+            {
+                // Try to reconnect
+                _reconnectionTryCount++;
+
+                if (_reconnectionTryCount <= MAXIMUM_RECONNECTION_ATTEMPTS)
+                {
+                    _client.Settings.useSecureProtocol = true;
+                    var t = _client.Reconnect(_room.Id, _room.SessionId);
+
+                    yield return new WaitUntil(() => t.IsCompleted);
+
+                    try
+                    {
+                        room = t.Result;
+
+                        Logger.Log($"Succesfully reconnected.");
+
+                        _room = room;
+                        AssignRoomCallbacks();
+                        SetConnectionState(ConnectionState.Connected);
+                        yield break;
+                    }
+                    catch (Exception e)
+                    {
+                        Logger.LogWarning($"Failed to reconnect - exception: {e.Message}");
+                    }
```

Comment:
curious why there's a `try/catch` block here, what bit of code could throw an error in the `try` block?

Reply:
The reconnecting Task `t` can fail here if no internet is available. I added that try and catch since that seemed like the best way to account for a Task failing to complete in C# and ensuring we don't run any "connection succesful" functions after a failed reconnect